### PR TITLE
sync attribute added to theano function

### PR DESCRIPTION
**Pull request by ReyhaneAskari**
Regarding ticket https://github.com/Theano/Theano/issues/6160, I added a sync attribute to theano function. I still need to figure out how to sync the outputs in addition to the shared variables.

### CODE REVIEW

File: theano/compile/tests/test_function_module.py [hunk 1]
```diff
@@ -907,6 +908,34 @@ def test_empty_givens_updates():
...
+
+        updates = [(w, w + T.dot(w, x) + T.dot(w, w))]
+
+        f = theano.function([x], y, updates=updates)
+        f.sync_shared()
+        g = theano.function([x], y, updates=updates)
+        x_ = np.random.rand(2000, 2000).astype('float32')
+        f(x_)
+        g(x_)
+        t_0 = time.time()
```
Comment (nouiz) [inline theano/compile/tests/test_function_module.py]: I don't understand this timing. I would do something like:

t0 = ..
for i in ...
    f()
    f.sync_update()
t1 = ...
for i in range ...
   f()
t2 = ...
assert t1-t0 is bigger then t2-t1.

the sync_update should slow down computation. I would use a small range then 1000. Maybe 10 or 50.

File: theano/compile/tests/test_function_module.py [hunk 2]
```diff
@@ -907,6 +908,34 @@ def test_empty_givens_updates():
     function([theano.In(x)], y, updates={})
 
 
+def test_sync():
+    if theano.config.device == 'cuda' and theano.gpuarray.pygpu_activated:
```
Comment (ReyhaneAskari) [inline theano/compile/tests/test_function_module.py]: Thanks. Done.

File: theano/compile/tests/test_function_module.py (hunk 1 above)
Comment (ReyhaneAskari) [inline theano/compile/tests/test_function_module.py]: The difference is not really significant, that's why I was using a large range. It's like 33.7922520638 and 33.9593539238. I think it's not being synced at all in this way. I'm thinking maybe we need to do this in the call and when we have access to the values. What to you think?

Comment (nouiz) [inline theano/compile/tests/test_function_module.py]: You need to call sync_shared() inside one of the loop. This is what will make a difference between both loop.

**Comment by ReyhaneAskari** [comment]
Thanks @nouiz ,with that change, the timing is 12.9660871029 and 13.467031002. Do you think that's significant enough?

**Comment by nouiz** [comment]
No. I'm actively working on it.

**Comment by nouiz** [comment]
I updated this PR with to have the test be meaningful. There was many problem with it. The function was only on CPU. The shared var was also on CPU. The function had an output, as the sync don't sync it, to easy the test I remove it. I made sure there is only 1 op that is inplace to don't have GPU memory allocation during the test.

I also modified how the timing is done to have the sync at the right place.

I also updated the timing to have CPU and GPU compute in parallel vs sequential in the other loop. This is the only way to get significant speed difference I found.

Also, I select the size dynamically to have the test more resistant to GPU changes. I tested it on a 750 and P100 SMX2.

**Pull request merged**